File: python/data_structures/hashtable.py
class HashTable:
	class Node:
		def __init__(self,key,value):
			self.key = key
			self.value = value

		def __str__(self):
			return f"Node({self.key},{self.value})"
		def __repr__(self):
			return self.__str__()

	def __str__(self):
		return str(self._buckets)


	def __init__(self,num_buckets = 263):
		self._num_buckets = num_buckets
		self._buckets = [ [] for _ in range(num_buckets)]

	def _hash(self,key):
		prime = 1000000007
		multiplier = 263
		hash = 0
		for x in key:
			hash = (hash * multiplier + ord(x)) % prime
		return hash % self._num_buckets

	def set(self, key, value):
		if key is not str:
			key = str(key)
		index = self._hash(key)
		for node in self._buckets[index]:
			if node.key == key:
				node.value = value
				break
		else:
			self._buckets[index].append(self.Node(key,value))

	def get(self, key):
		if key is not str:
			key = str(key)
		idx = self._hash(key)
		if not self._buckets[idx]:
			return None
		else:
			for node in self._buckets[idx]:
				if node.key == key:
					return node.value
			else:
				return None


def main():
	ht = HashTable(20)
	ht.set('b',100)
	ht.set('a',30)
	ht.set('c',40)
	print(ht)
	print(ht.get('a'))

if __name__ == "__main__":
	main()

File: js/data_structures/hashtable.js
class HashTable{
	static Node = class{
		constructor(key,value){
			this.key = key;
			this.value = value;
		}
		toString(){
			return `{${this.key}:${this.value}}`;
		}
	}
	constructor(num_buckets = 263){
		this._num_buckets = num_buckets;
		this._buckets = new Array(num_buckets);
	}
	_hash(key){
		let prime = 1000000007;
		let multiplier = 263;
		let hash = 0;
		for(let x in key){
			hash = (hash * multiplier + x.charCodeAt(0)) % prime;
		}
		return hash % this._num_buckets;
	}

	set(key,value){
		if(typeof key !== "string")
			key = key.toString();
		let index = this._hash(key);
		if(this._buckets[index] === undefined){
			this._buckets[index] = [new HashTable.Node(key,value)];
			return;
		}
		for(let i = 0;i< this._num_buckets[index];i++){
			let node = this._buckets[i];
			if(node.key === key){
				this._buckets[i].value = value;
				return;
			}
		}
		this._buckets[index].push(new HashTable.Node(key,value));
	}
	get(key){
		if(typeof key !== "string")
			key = key.toString();

		let index = this._hash(key)
		if(!this._buckets[index])
			return null;
		else{
			for(let i=0;i<this._num_buckets;i++){
				let node = this._buckets[i];
				if(node.key === key)
					return node.value;
			}
			return null;
		}
	}
	toString(){
		let os = '{';
		for(let i = 0;i<this._num_buckets;i++){
			if(this._buckets[i] === undefined)
				continue;
			for(let j = 0; j<this._buckets[i].length;j++){
				let node = this._buckets[i][j];
				os+= `${node.key}:${node.value}`;
				if(j < this._buckets[i].length - 1){
					os+= ', ';
				}
			}
		}
		os+='}';
		return os;
	}
}

if(typeof module !== 'undefined' && require.main === module){
	let ht = new HashTable();
	ht.set('a',3);
	ht.set('b',4);
	ht.set('c',5);
	ht.set('d',6);
	ht.set('e',7);
	console.log(ht.toString());
}